File: packages/ts-host/src/mcp/mcp-server.ts
import type { Request, Response } from 'express'
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
import type { AgentContext } from '@ai37/agent-sdk'
import { currentCtx } from '../als'
import type { McpOptions, McpToolDef } from './types'

/**
 * Собирает `McpServer` из набора tool-определений. `ctx` (кто вызвал) прокидывается в каждый
 * handler. `@modelcontextprotocol/sdk` и `zod` подгружаются ДИНАМИЧЕСКИ — это делает их
 * optional-peer'ами: потребители agent-host, не использующие `mcp`, их не тянут. Без ALS —
 * чтобы переиспользовать и вне agent-host (напр. в NestJS rag-factory со своим guard'ом).
 */
export async function buildMcpServer(
  serverInfo: { name: string; version: string },
  tools: McpToolDef[],
  ctx: AgentContext | undefined,
): Promise<McpServer> {
  const [{ McpServer: McpServerCtor }, { z }] = await Promise.all([
    import('@modelcontextprotocol/sdk/server/mcp.js'),
    import('zod'),
  ])
  const defaultInputSchema = {
    query: z.string().describe('Запрос на естественном языке'),
  }
  const server = new McpServerCtor({
    name: serverInfo.name,
    version: serverInfo.version,
  })
  for (const t of tools) {
    server.registerTool(
      t.name,
      {
        description: t.description,
        inputSchema: t.inputSchema ?? defaultInputSchema,
      },
      async (args: Record<string, unknown>) => {
        const result = await t.handler(args, ctx)
        // McpToolResult — подмножество CallToolResult; форма совпадает.
        return result as Awaited<
          ReturnType<Parameters<McpServer['registerTool']>[2]>
        >
      },
    )
  }
  return server
}

/**
 * Express-handler MCP-эндпоинта (StreamableHTTP, stateless: новый server+transport на запрос —
 * это и позволяет chat-backend отдавать НАБОР ПО ПОЛЬЗОВАТЕЛЮ). `ctx` берётся из ALS
 * (`mcpChallengeGuard` его туда положил), tools резолвятся статически или per-request.
 */
export function mcpHttpHandler(
  opts: McpOptions,
  serverInfo: { name: string; version: string },
) {
  return async (req: Request, res: Response): Promise<void> => {
    const ctx = currentCtx()
    const resolved =
      typeof opts.tools === 'function' ? await opts.tools(ctx) : opts.tools
    // Резолвер мог вернуть либо список, либо { tools, release } (занял ресурсы на запрос).
    const tools = Array.isArray(resolved) ? resolved : resolved.tools
    const release = Array.isArray(resolved) ? undefined : resolved.release

    const server = await buildMcpServer(serverInfo, tools, ctx)
    const { StreamableHTTPServerTransport } = await import(
      '@modelcontextprotocol/sdk/server/streamableHttp.js'
    )
    // Stateless: без sessionId (сервер-инициированный SSE не нужен, запросы независимы).
    const transport = new StreamableHTTPServerTransport({
      sessionIdGenerator: undefined,
    })
    res.on('close', () => {
      void transport.close()
      void server.close()
      if (release) void release()
    })
    await server.connect(transport)
    // express.json() уже распарсил тело → передаём его третьим аргументом.
    await transport.handleRequest(req, res, req.body)
  }
}

File: packages/python-host/src/ai37_agent_host/mcp/mcp_server.py
"""MCP-сервер поверх официального python ``mcp`` SDK — порт ``ts-host/src/mcp/mcp-server.ts``.

Собирает low-level ``mcp.server.lowlevel.Server`` из набора tool-определений и отдаёт его как
ASGI-приложение (StreamableHTTP, stateless). ``mcp`` SDK — soft-import (optional-группа ``mcp``):
потребители host'а, не использующие MCP-экспорт, его не тянут. Отсутствие пакета → понятная
ошибка при попытке смонтировать MCP (см. :class:`MissingMcpDependencyError`).

Расхождение с TS (``@modelcontextprotocol/sdk``), для ревью:
  * TS создаёт НОВЫЙ ``McpServer``+``StreamableHTTPServerTransport`` НА КАЖДЫЙ HTTP-запрос
    (полностью stateless, session-less). Python-SDK устроен иначе: ``Server`` долгоживущий, а
    ``StreamableHTTPSessionManager`` держит сессии; per-request build не ложится на его модель.
  * ЭКВИВАЛЕНТ per-user набора здесь достигается тем, что ``list_tools``/``call_tool``
    low-level ``Server`` вызываются НА КАЖДЫЙ запрос — резолвер tools (в т.ч. per-user из ALS)
    исполняется внутри них. Так же, как в TS, ``ctx`` берётся из ALS (challenge-guard положил).
  * ``release`` резолвера вызывается в ``finally`` вокруг обработки одного tool-вызова
    (в TS — на ``res.on('close')``); семантика «занял ресурсы на запрос → освободи» сохранена.
"""

from __future__ import annotations

from collections.abc import Awaitable
from dataclasses import dataclass
from typing import Any

from ..als import current_ctx
from .types import McpOptions, McpToolDef, McpToolResult, McpToolSet

_DEFAULT_INPUT_SCHEMA: dict[str, Any] = {
    "type": "object",
    "properties": {
        "query": {"type": "string", "description": "Запрос на естественном языке"},
    },
    "required": ["query"],
}


class MissingMcpDependencyError(RuntimeError):
    """MCP-экспорт запрошен, но пакет ``mcp`` не установлен (optional-группа ``mcp``)."""

    def __init__(self) -> None:
        super().__init__(
            "MCP-экспорт требует пакет 'mcp' (официальный Model Context Protocol Python SDK). "
            "Установите optional-группу: poetry install --with mcp (или pip install mcp)."
        )


def _import_mcp() -> Any:
    """Soft-import ``mcp`` SDK. Бросает :class:`MissingMcpDependencyError`, если пакета нет."""
    try:
        import mcp.types as mcp_types  # noqa: F401
        from mcp.server.lowlevel import Server  # noqa: F401
    except ImportError as exc:  # pragma: no cover - зависит от установки optional-группы
        raise MissingMcpDependencyError() from exc
    return mcp_types, Server


@dataclass
class ServerInfo:
    """Идентификация MCP-сервера в ``initialize`` (имя + версия)."""

    name: str
    version: str


async def _resolve_tools(
    opts: McpOptions,
    ctx: Any,
) -> tuple[list[McpToolDef], Any]:
    """Резолвит tools (статический список или per-request функцию) → ``(tools, release)``.

    Функция-резолвер может вернуть список ИЛИ :class:`McpToolSet` (занял ресурсы на запрос),
    sync или async. Зеркало ветвления ``mcpHttpHandler`` в TS.
    """
    resolver = opts.tools
    if callable(resolver):
        resolved = resolver(ctx)
        if isinstance(resolved, Awaitable):
            resolved = await resolved
    else:
        resolved = resolver

    if isinstance(resolved, McpToolSet):
        return list(resolved.tools), resolved.release
    return list(resolved), None


async def _run_tool(tool: McpToolDef, args: dict[str, Any], ctx: Any) -> McpToolResult:
    """Вызвать handler tool'а (sync или async) с verified ``ctx``."""
    result = tool.handler(args, ctx)
    if isinstance(result, Awaitable):
        result = await result
    return result


async def _call_release(release: Any) -> None:
    """Освободить ресурсы резолвера (sync или async), проглатывая ошибки."""
    if release is None:
        return
    try:
        outcome = release()
        if isinstance(outcome, Awaitable):
            await outcome
    except Exception:  # noqa: BLE001 - освобождение не должно ронять ответ
        pass


def build_mcp_server(server_info: ServerInfo, opts: McpOptions) -> Any:
    """Собрать low-level ``mcp`` ``Server`` с per-request резолвом tools.

    ``list_tools``/``call_tool`` резолвят набор на каждый запрос (читая ``ctx`` из ALS) — так
    достигается per-user набор. Возвращает объект ``mcp.server.lowlevel.Server`` (типизирован
    ``Any``, чтобы host не тянул ``mcp`` в обязательные deps).
    """
    mcp_types, Server = _import_mcp()
    server = Server(server_info.name, version=server_info.version)

    @server.list_tools()
    async def _list_tools() -> list[Any]:  # pyright: ignore[reportUnusedFunction]
        ctx = current_ctx()
        tools, release = await _resolve_tools(opts, ctx)
        try:
            return [
                mcp_types.Tool(
                    name=t.name,
                    description=t.description,
                    inputSchema=t.input_schema or _DEFAULT_INPUT_SCHEMA,
                )
                for t in tools
            ]
        finally:
            await _call_release(release)

    @server.call_tool()
    async def _call_tool(name: str, arguments: dict[str, Any]) -> list[Any]:  # pyright: ignore[reportUnusedFunction]
        ctx = current_ctx()
        tools, release = await _resolve_tools(opts, ctx)
        try:
            tool = next((t for t in tools if t.name == name), None)
            if tool is None:
                raise ValueError(f"Unknown tool: {name}")
            result = await _run_tool(tool, arguments or {}, ctx)
            content = [
                mcp_types.TextContent(type="text", text=block.get("text", ""))
                for block in result.content
            ]
            # low-level call_tool: (content, structuredContent) — второй элемент нам не нужен.
            return content
        finally:
            await _call_release(release)

    return server


def create_mcp_asgi_app(server_info: ServerInfo, opts: McpOptions) -> tuple[Any, Any]:
    """ASGI-приложение MCP-эндпоинта (StreamableHTTP, stateless) + его session-manager.

    Возвращает ``(asgi_app, session_manager)``. ``asgi_app`` монтируется на ``/mcp`` хоста;
    ``session_manager`` ОБЯЗАН быть запущен в lifespan top-level приложения
    (``async with session_manager.run(): ...``) — см. модуль-docstring и ``mount.py``.
    """
    try:
        from mcp.server.streamable_http_manager import StreamableHTTPSessionManager
    except ImportError as exc:  # pragma: no cover - зависит от установки optional-группы
        raise MissingMcpDependencyError() from exc

    server = build_mcp_server(server_info, opts)
    # stateless=True: без Mcp-Session-Id, каждый запрос независим (зеркало TS sessionId=undefined).
    session_manager = StreamableHTTPSessionManager(app=server, stateless=True, json_response=True)

    async def asgi_app(scope: dict[str, Any], receive: Any, send: Any) -> None:
        await session_manager.handle_request(scope, receive, send)

    return asgi_app, session_manager
